text/0000-read-only-build-layers.md

```diff
@@ -0,0 +1,98 @@
+
+## What use cases does it support?
+
+- Allowing buildpacks to explicitly indicate how the layers provided by them should be used and having this be guaranteed by the lifecycle.
+
+### Example use case for read-only build layers
+
+A build process consists of 2 buildpacks - a python buildpack and a pip buildpack
+
+1. Buildpack #1 provides a python binary that can be used by subsequent buildpacks but is also used in the app image
+2. Buildpack #2 uses the python binary to invoke pip to provide further dependencies.
+
+Unbeknownst to the buildpack author, python has a side-effect on execution where if it's installation directory is writable, it creates new cache objects or updates existing files. As a result, execution of buildpack #2 causes the content of the layers provided by buildpack #1 to change, causing the final layer imageID to change and hampers layer reproducibility and new layers have to be pushed out to the registry.
+
+This could easily be fixed if buildpack #1 only exposed its build layers as read-only to subsequent buildpacks or if it discarded the modifications made by subsequent buildpacks essentially making it "read-only" in the context of the build process.
+
+## What is the expected outcome?
+
+The spec and lifecycle is modified to support the above use cases.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Updates to the lifecycle so that `build` layers created by buildpacks should be layerized immediately after the buildpack that provided the layer is finished with its build process. This way any modification made by the future buildpacks would be ignore while still making this layer available to future buildpacks.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The lifecycle will have to layerize the layers marked `build` as `true` during the `build` phase of the lifecycle instead of the `export` phase.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Possibly mixing the concerns of the `build` and the `export` phases.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered?
+
+We could change the the layer so that it is owned by root or a different user between the build steps of different buildpacks so that subsequent buildpacks cannot modify the layers created by other buildpacks. This would require elevated privileges during the build phase of the lifecycle.
+
+
+- Why is this proposal the best?
+
+This proposal allows you to achieve the above use cases without changing the expected UID/GID of the layers and without elevated permissions during the build phase.
+
+- What is the impact of not doing this?
+
+We could have unknown layer modifications by buildpacks which is currently a spec violation.
+
+# Prior Art
+[prior-art]: #prior-art
+
+TODO
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+The impact this would have on the performance of the export step. The exact details of which parts of the lifecycle build and export phases that need to be changed.
```

Comment:
Given the process of archiving layers in currently sequential, if we simply moving layer archiving from the export to the build phase the performance of a the build operation as a whole will not change. However, the duration of the build phase will increase while the duration of the export phase will decrease. Interleaving the slow of archiving and calculating diff IDs between the execution of individuals buildpacks may impact the perception users have of the performance of the system. Also it will impact the time of an iteration loop on a failed build

Reply:
I worry about this. We (Paketo Buildpacks) sometimes get reports that a buildpack we maintain is slow to execute when the user is really reporting that parts of the restore or export phases are slow. I'd like to reduce those types of reports if at all possible and I'm concerned that if we are doing this "layerizing" in the build phase we may see an increase in reports that a buildpack is slow when it is really that the "layerizing" process is taking a while to perform its operation.

Reply:
We could layerize previous buildpack layers (non-`build=true`-only) in parallel with the rest of the build process. This would result in net faster builds.